File: benchmark/2026-09-17-7677581(latest)/micro/reference/ref.c
/* Reference implementation of the 11 frozen micro-benchmark workloads.
   Used ONLY to derive the expected output values recorded in the frozen methodology. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef long long i64;
static const i64 P = 2147483647LL;
static const i64 MA = 48271LL;
static i64 st;
static void rseed(i64 s) { st = s; }
static i64 nexti(void) { st = (MA * st) % P; return st; }
static double nextu(void) { return (double)nexti() / 2147483647.0; }

static void pf(const char *n, double v) { printf(" %s=%.16e", n, v); }

/* ---------- MB01 ---------- */
static i64 fibr(int n) { if (n < 2) return n; return fibr(n - 1) + fibr(n - 2); }
static void mb01(void) {
    i64 t = 0;
    for (int n = 30; n <= 37; n++) t += fibr(n);
    printf("MB01 %lld\n", t);
}

/* ---------- MB02 ---------- */
static void mb02(void) {
    const i64 M = 1000003, N = 20000;
    i64 total = 0;
    for (i64 k = 1; k <= N; k++) {
        i64 f = 1;
        for (i64 j = 2; j <= k; j++) f = (f * j) % M;
        total = (total + f) % M;
    }
    printf("MB02 %lld\n", total);
}

/* ---------- MB03 ---------- */
static void mb03(void) {
    const i64 N = 120000000LL;
    rseed(20263917LL);
    i64 x = st, s_add = 0, s_xor = 0, s_mul = 1, s_div = 0;
    for (i64 i = 0; i < N; i++) {
        x = (48271LL * x) % 2147483647LL;
        s_add = (s_add + x) % 2147483647LL;
        s_xor = s_xor ^ x;
        s_mul = (s_mul * 33 + (x % 97)) % 1000003LL;
        s_div = s_div + (x / 1000);
    }
    printf("MB03 %lld %lld %lld %lld\n", s_add, s_xor, s_mul, s_div);
}

/* ---------- MB04 ---------- */
static void mb04(void) {
    const int M = 4000; const int R = 75000;
    static double A[4000], B[4000];
    rseed(20264917LL);
    for (int i = 0; i < M; i++) A[i] = 0.5 + nextu();
    for (int i = 0; i < M; i++) B[i] = 0.5 + nextu();
    double s1 = 0, s2 = 0, s3 = 0, s4 = 0;
    for (int r = 0; r < R; r++) {
        A[r % M] = A[r % M] + 1.0e-9;
        for (int i = 0; i < M; i++) {
            double a = A[i], b = B[i];
            s1 = s1 + a * b;
            s2 = s2 + a / (b + 2.0);
            s3 = s3 + sqrt(a * a + b * b);
            s4 = s4 + (a - b) * (a - b);
        }
    }
    printf("MB04"); pf("s1", s1); pf("s2", s2); pf("s3", s3); pf("s4", s4); printf("\n");
}

/* ---------- MB05 ---------- */
static void mb05(void) {
    const int N = 2000000; const int R = 400;
    double *X = malloc(sizeof(double) * N), *Y = malloc(sizeof(double) * N);
    rseed(20265917LL);
    for (int i = 0; i < N; i++) X[i] = 0.5 + nextu();
    for (int i = 0; i < N; i++) Y[i] = 0.5 + nextu();
    double total = 0.0;
    for (int r = 0; r < R; r++) {
        X[r] = X[r] + 1.0e-9;
        double d = 0.0;
        for (int i = 0; i < N; i++) d = d + X[i] * Y[i];
        total = total + d;
    }
    printf("MB05"); pf("total", total); printf("\n");
    free(X); free(Y);
}

/* ---------- MB06 ---------- */
static void mb06(void) {
    const int n = 512; const int R = 3;
    double *A = malloc(sizeof(double) * n * n), *B = malloc(sizeof(double) * n * n), *C = malloc(sizeof(double) * n * n);
    rseed(20266917LL);
    for (int i = 0; i < n * n; i++) A[i] = nextu();
    for (int i = 0; i < n * n; i++) B[i] = nextu();
    for (int r = 0; r < R; r++) {
        A[r] = A[r] + 1.0e-9;
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++) {
                double s = 0.0;
                for (int k = 0; k < n; k++) s = s + A[i * n + k] * B[k * n + j];
                C[i * n + j] = s;
            }
    }
    double sum = 0.0;
    for (int i = 0; i < n * n; i++) sum = sum + C[i];
    printf("MB06"); pf("sumC", sum); pf("c_first", C[0]); pf("c_last", C[n * n - 1]); printf("\n");
    free(A); free(B); free(C);
}

/* ---------- MB07 ---------- */
static void msort(i64 *a, i64 *buf, int n) {
    i64 *src = a, *dst = buf;
    for (int width = 1; width < n; width *= 2) {
        for (int lo = 0; lo < n; lo += 2 * width) {
            int mid = lo + width; if (mid > n) mid = n;
            int hi = lo + 2 * width; if (hi > n) hi = n;
            int i = lo, j = mid, k = lo;
            while (i < mid && j < hi) { if (src[i] <= src[j]) dst[k++] = src[i++]; else dst[k++] = src[j++]; }
            while (i < mid) dst[k++] = src[i++];
            while (j < hi) dst[k++] = src[j++];
        }
        i64 *t = src; src = dst; dst = t;
    }
    if (src != a) for (int i = 0; i < n; i++) a[i] = src[i];
}
static void mb07(void) {
    const int N = 2000000; const int R = 4;
    i64 *src = malloc(sizeof(i64) * N), *a = malloc(sizeof(i64) * N), *buf = malloc(sizeof(i64) * N);
    rseed(20267917LL);
    for (int i = 0; i < N; i++) src[i] = nexti();
    i64 total = 0, sum = 0, inv = 0;
    for (int r = 0; r < R; r++) {
        src[r] = src[r] + 1;
        for (int i = 0; i < N; i++) a[i] = src[i];
        msort(a, buf, N);
        i64 chk = 0;
        for (int i = 0; i < N; i++) chk = (chk * 31 + (a[i] % 1000003LL)) % 1000003LL;
        total = (total * 7 + chk) % 1000003LL;
        sum = 0;
        for (int i = 0; i < N; i++) sum += a[i];
        for (int i = 1; i < N; i++) if (a[i - 1] > a[i]) inv++;
    }
    printf("MB07 %lld %lld %lld\n", total, sum, inv);
    free(src); free(a); free(buf);
}

/* ---------- MB08 ---------- */
static i64 rhash(const unsigned char *s, i64 len) {
    i64 h = 0;
    for (i64 i = 0; i < len; i++) h = (h * 131 + (i64)s[i]) % 1000000007LL;
    return h;
}
static void mb08(void) {
    const int NW = 200000; const int R = 20;
    unsigned char *text = malloc(NW * 20 + 16);
    i64 len = 0;
    rseed(20268917LL);
    for (int w = 0; w < NW; w++) {
        if (w > 0) text[len++] = ' ';
        int L = 4 + (int)(nexti() % 13);
        for (int c = 0; c < L; c++) text[len++] = (unsigned char)('a' + (int)(nexti() % 26));
    }
    unsigned char *U = malloc(len + 1), *Rv = malloc(len + 1);
    i64 acc = 0, cnt_ab = 0, cnt_w = 0;
    const i64 Q = 1000000007LL;
    for (int r = 0; r < R; r++) {
        i64 p = 7LL * r + 11;
        if (text[p] == ' ') text[p] = 'x';
        else text[p] = (unsigned char)('a' + ((text[p] - 'a' + 1) % 26));
        i64 h1 = rhash(text, len);
        for (i64 i = 0; i < len; i++) { unsigned char c = text[i]; U[i] = (c >= 'a' && c <= 'z') ? (unsigned char)(c - 32) : c; }
        i64 h2 = rhash(U, len);
        for (i64 i = 0; i < len; i++) Rv[i] = text[len - 1 - i];
        i64 h3 = rhash(Rv, len);
        cnt_ab = 0;
        for (i64 i = 0; i + 1 < len; i++) if (text[i] == 'a' && text[i + 1] == 'b') cnt_ab++;
        cnt_w = 1;
        for (i64 i = 0; i < len; i++) if (text[i] == ' ') cnt_w++;
        acc = (acc * 31 + h1) % Q;
        acc = (acc * 31 + h2) % Q;
        acc = (acc * 31 + h3) % Q;
        acc = (acc * 31 + cnt_ab) % Q;
        acc = (acc * 31 + cnt_w) % Q;
    }
    printf("MB08 %lld %lld %lld %lld\n", acc, len, cnt_ab, cnt_w);
    free(text); free(U); free(Rv);
}

/* ---------- MB09 ---------- */
static void mb09(void) {
    const int N = 2000000; const int R = 30;
    double *X = malloc(sizeof(double) * N), *Y = malloc(sizeof(double) * N);
    rseed(20269917LL);
    for (int i = 0; i < N; i++) X[i] = nextu() * 100.0;
    for (int i = 0; i < N; i++) Y[i] = nextu() * 100.0;
    double mean = 0, var = 0, sd = 0, mn = 0, mx = 0, mad = 0, pear = 0;
    i64 hchk = 0;
    for (int r = 0; r < R; r++) {
        X[r] = X[r] + 1.0e-9;
        double sum = 0.0; mn = X[0]; mx = X[0];
        for (int i = 0; i < N; i++) { double v = X[i]; sum = sum + v; if (v < mn) mn = v; if (v > mx) mx = v; }
        mean = sum / (double)N;
        double sq = 0.0, ad = 0.0;
        for (int i = 0; i < N; i++) { double d = X[i] - mean; sq = sq + d * d; ad = ad + (d < 0 ? -d : d); }
        var = sq / (double)N; sd = sqrt(var); mad = ad / (double)N;
        double sy = 0.0;
        for (int i = 0; i < N; i++) sy = sy + Y[i];
        double meany = sy / (double)N;
        double sxy = 0.0, sxx = 0.0, syy = 0.0;
        for (int i = 0; i < N; i++) { double dx = X[i] - mean, dy = Y[i] - meany; sxy = sxy + dx * dy; sxx = sxx + dx * dx; syy = syy + dy * dy; }
        pear = sxy / sqrt(sxx * syy);
        i64 hist[64]; for (int b = 0; b < 64; b++) hist[b] = 0;
        for (int i = 0; i < N; i++) {
            int b = (int)floor(X[i] * 0.64);
            if (b < 0) b = 0; if (b > 63) b = 63;
            hist[b]++;
        }
        hchk = 0;
        for (int b = 0; b < 64; b++) hchk += (i64)(b + 1) * hist[b];
    }
    printf("MB09"); pf("mean", mean); pf("var", var); pf("sd", sd); pf("min", mn); pf("max", mx);
    pf("mad", mad); pf("pearson", pear); printf(" hist_chk=%lld\n", hchk);
    free(X); free(Y);
}

/* ---------- MB10 ---------- */
static void mb10(void) {
    const int N = 1000000; const int R = 3;
    rseed(20270917LL);
    i64 sum_v = 0, chk = 0, bytes = 0, lines = 0;
    char name[64], line[128];
    for (int r = 0; r < R; r++) {
        snprintf(name, sizeof name, "mb10_round_%d.txt", r);
        FILE *f = fopen(name, "wb");
        for (int i = 0; i < N; i++) {
            i64 v = nexti();
            int n = snprintf(line, sizeof line, "%d %lld\n", i, v);
            fwrite(line, 1, (size_t)n, f);
            bytes += n;
        }
        fclose(f);
        f = fopen(name, "rb");
        char buf[256];
        i64 idx = 0;
        while (fgets(buf, sizeof buf, f)) {
            char *e;
            i64 a = strtoll(buf, &e, 10);
            i64 v = strtoll(e, &e, 10);
            if (a != idx) { fprintf(stderr, "MB10 index mismatch\n"); exit(1); }
            idx++; lines++;
            sum_v = (sum_v + v) % 1000000007LL;
            chk = (chk * 31 + (v % 1000003LL)) % 1000003LL;
        }
        fclose(f);
    }
    printf("MB10 %lld %lld %lld %lld\n", sum_v, chk, bytes, lines);
}

/* ---------- MB11 ---------- */
static void mb11(void) {
    const int N = 1000000; const int R = 3;
    const int KM = 500009, SM = 100003;
    const i64 Q = 1000000007LL;
    i64 *keys = malloc(sizeof(i64) * N);
    rseed(20271917LL);
    for (int i = 0; i < N; i++) keys[i] = nexti();
    i64 *cnt = malloc(sizeof(i64) * KM);
    char *present = malloc(KM);
    char *sset = malloc(SM);
    i64 acc = 0, size1 = 0, found = 0, vsum = 0, mchk = 0, size2 = 0, size3 = 0, lsum = 0;
    for (int r = 0; r < R; r++) {
        keys[r] = keys[r] + 1000000;
        memset(cnt, 0, sizeof(i64) * KM); memset(present, 0, KM);
        size1 = 0;
        for (int i = 0; i < N; i++) {
            i64 k = keys[i] % KM;
            if (!present[k]) { present[k] = 1; size1++; cnt[k] = 1; } else cnt[k] += 1;
        }
        found = 0; vsum = 0;
        for (int i = 0; i < N; i++) {
            i64 k = (keys[i] + 7) % KM;
            if (present[k]) { found++; vsum += cnt[k]; }
        }
        mchk = 0;
        for (i64 k = 0; k < KM; k++) if (present[k]) mchk = (mchk + (k % 1000003LL) * cnt[k]) % 1000003LL;
        for (int i = 0; i < N; i += 2) {
            i64 k = keys[i] % KM;
            if (present[k]) { present[k] = 0; cnt[k] = 0; }
        }
        size2 = 0;
        for (i64 k = 0; k < KM; k++) if (present[k]) size2++;
        memset(sset, 0, SM);
        size3 = 0;
        for (int i = 0; i < N; i++) { i64 k = keys[i] % SM; if (!sset[k]) { sset[k] = 1; size3++; } }
        lsum = 0;
        for (int i = 0; i < N; i++) lsum += keys[i] % 1000;
        acc = (acc * 31 + size1) % Q;
        acc = (acc * 31 + found) % Q;
        acc = (acc * 31 + vsum) % Q;
        acc = (acc * 31 + mchk) % Q;
        acc = (acc * 31 + size2) % Q;
        acc = (acc * 31 + size3) % Q;
        acc = (acc * 31 + lsum) % Q;
    }
    printf("MB11 %lld %lld %lld %lld %lld %lld %lld %lld\n", acc, size1, found, vsum, mchk, size2, size3, lsum);
    free(keys); free(cnt); free(present); free(sset);
}

int main(int argc, char **argv) {
    const char *only = argc > 1 ? argv[1] : "all";
    #define RUN(tag, fn) if (!strcmp(only, "all") || !strcmp(only, tag)) { fn(); fflush(stdout); }
    RUN("mb01", mb01) RUN("mb02", mb02) RUN("mb03", mb03) RUN("mb04", mb04)
    RUN("mb05", mb05) RUN("mb06", mb06) RUN("mb07", mb07) RUN("mb08", mb08)
    RUN("mb09", mb09) RUN("mb10", mb10) RUN("mb11", mb11)
    return 0;
}

File: benchmark/template/fixtures/micro/reference/ref.py
#!/usr/bin/env python3
# Independent Python implementation of the 11 frozen micro workloads,
# written from the frozen prose spec, used to cross-validate the C reference.
import math, os, sys

P = 2147483647
A_MUL = 48271
MAXEXACT = 2**53
_bound = 0


def bound(v):
    global _bound
    if v > _bound:
        _bound = v
    return v


class R:
    def __init__(self, seed):
        self.s = seed

    def i(self):
        bound(A_MUL * self.s)
        self.s = (A_MUL * self.s) % P
        return self.s

    def u(self):
        return self.i() / 2147483647.0


def fmt(x):
    return "%.16e" % x


def mb01():
    sys.setrecursionlimit(100000)
    def fib(n):
        return n if n < 2 else fib(n - 1) + fib(n - 2)
    t = 0
    for n in range(30, 38):
        t += fib(n)
    print("MB01 %d" % t, flush=True)


def mb02():
    M, N = 1000003, 20000
    total = 0
    for k in range(1, N + 1):
        f = 1
        for j in range(2, k + 1):
            bound(f * j)
            f = (f * j) % M
        total = (total + f) % M
    print("MB02 %d" % total, flush=True)


def mb03():
    N = 120000000
    x = 20263917
    sa = sx = sd = 0
    sm = 1
    for _ in range(N):
        x = (48271 * x) % 2147483647
        sa = (sa + x) % 2147483647
        sx = sx ^ x
        sm = (sm * 33 + (x % 97)) % 1000003
        sd = sd + (x // 1000)
    bound(sd)
    print("MB03 %d %d %d %d" % (sa, sx, sm, sd), flush=True)


def mb04():
    M, RR = 4000, 75000
    g = R(20264917)
    A = [0.5 + g.u() for _ in range(M)]
    B = [0.5 + g.u() for _ in range(M)]
    s1 = s2 = s3 = s4 = 0.0
    for r in range(RR):
        A[r % M] = A[r % M] + 1.0e-9
        for i in range(M):
            a = A[i]; b = B[i]
            s1 = s1 + a * b
            s2 = s2 + a / (b + 2.0)
            s3 = s3 + math.sqrt(a * a + b * b)
            s4 = s4 + (a - b) * (a - b)
    print("MB04 s1=%s s2=%s s3=%s s4=%s" % (fmt(s1), fmt(s2), fmt(s3), fmt(s4)), flush=True)


def mb05():
    N, RR = 2000000, 400
    g = R(20265917)
    X = [0.5 + g.u() for _ in range(N)]
    Y = [0.5 + g.u() for _ in range(N)]
    total = 0.0
    for r in range(RR):
        X[r] = X[r] + 1.0e-9
        d = 0.0
        for i in range(N):
            d = d + X[i] * Y[i]
        total = total + d
    print("MB05 total=%s" % fmt(total), flush=True)


def mb06():
    n, RR = 512, 3
    g = R(20266917)
    A = [g.u() for _ in range(n * n)]
    B = [g.u() for _ in range(n * n)]
    C = [0.0] * (n * n)
    for r in range(RR):
        A[r] = A[r] + 1.0e-9
        for i in range(n):
            for j in range(n):
                s = 0.0
                for k in range(n):
                    s = s + A[i * n + k] * B[k * n + j]
                C[i * n + j] = s
    tot = 0.0
    for v in C:
        tot = tot + v
    print("MB06 sumC=%s c_first=%s c_last=%s" % (fmt(tot), fmt(C[0]), fmt(C[n * n - 1])), flush=True)


def msort(a, buf, n):
    src, dst = a, buf
    width = 1
    while width < n:
        lo = 0
        while lo < n:
            mid = min(lo + width, n)
            hi = min(lo + 2 * width, n)
            i, j, k = lo, mid, lo
            while i < mid and j < hi:
                if src[i] <= src[j]:
                    dst[k] = src[i]; i += 1
                else:
                    dst[k] = src[j]; j += 1
                k += 1
            while i < mid:
                dst[k] = src[i]; i += 1; k += 1
            while j < hi:
                dst[k] = src[j]; j += 1; k += 1
            lo += 2 * width
        src, dst = dst, src
        width *= 2
    if src is not a:
        for i in range(n):
            a[i] = src[i]


def mb07():
    N, RR = 2000000, 4
    g = R(20267917)
    src = [g.i() for _ in range(N)]
    buf = [0] * N
    total = 0; ssum = 0; inv = 0
    for r in range(RR):
        src[r] = src[r] + 1
        a = list(src)
        msort(a, buf, N)
        chk = 0
        for i in range(N):
            bound(chk * 31)
            chk = (chk * 31 + (a[i] % 1000003)) % 1000003
        total = (total * 7 + chk) % 1000003
        ssum = 0
        for v in a:
            ssum += v
        bound(ssum)
        for i in range(1, N):
            if a[i - 1] > a[i]:
                inv += 1
    print("MB07 %d %d %d" % (total, ssum, inv), flush=True)


def rhash(s):
    h = 0
    for c in s:
        bound(h * 131 + c)
        h = (h * 131 + c) % 1000000007
    return h


def mb08():
    NW, RR = 200000, 20
    g = R(20268917)
    parts = []
    for w in range(NW):
        L = 4 + (g.i() % 13)
        parts.append(bytes(97 + (g.i() % 26) for _ in range(L)))
    text = bytearray(b" ".join(parts))
    ln = len(text)
    Q = 1000000007
    acc = 0; cnt_ab = 0; cnt_w = 0
    for r in range(RR):
        p = 7 * r + 11
        if text[p] == 32:
            text[p] = ord('x')
        else:
            text[p] = 97 + ((text[p] - 97 + 1) % 26)
        h1 = rhash(text)
        U = bytearray(ln)
        for i in range(ln):
            c = text[i]
            U[i] = c - 32 if 97 <= c <= 122 else c
        h2 = rhash(U)
        Rv = bytearray(ln)
        for i in range(ln):
            Rv[i] = text[ln - 1 - i]
        h3 = rhash(Rv)
        cnt_ab = 0
        for i in range(ln - 1):
            if text[i] == 97 and text[i + 1] == 98:
                cnt_ab += 1
        cnt_w = 1
        for i in range(ln):
            if text[i] == 32:
                cnt_w += 1
        for v in (h1, h2, h3, cnt_ab, cnt_w):
            bound(acc * 31 + v)
            acc = (acc * 31 + v) % Q
    print("MB08 %d %d %d %d" % (acc, ln, cnt_ab, cnt_w), flush=True)


def mb09():
    N, RR = 2000000, 30
    g = R(20269917)
    X = [g.u() * 100.0 for _ in range(N)]
    Y = [g.u() * 100.0 for _ in range(N)]
    for r in range(RR):
        X[r] = X[r] + 1.0e-9
        s = 0.0; mn = X[0]; mx = X[0]
        for v in X:
            s = s + v
            if v < mn: mn = v
            if v > mx: mx = v
        mean = s / N
        sq = 0.0; ad = 0.0
        for v in X:
            d = v - mean
            sq = sq + d * d
            ad = ad + (-d if d < 0 else d)
        var = sq / N; sd = math.sqrt(var); mad = ad / N
        sy = 0.0
        for v in Y:
            sy = sy + v
        meany = sy / N
        sxy = 0.0; sxx = 0.0; syy = 0.0
        for i in range(N):
            dx = X[i] - mean; dy = Y[i] - meany
            sxy = sxy + dx * dy; sxx = sxx + dx * dx; syy = syy + dy * dy
        pear = sxy / math.sqrt(sxx * syy)
        hist = [0] * 64
        for v in X:
            b = int(math.floor(v * 0.64))
            if b < 0: b = 0
            if b > 63: b = 63
            hist[b] += 1
        hchk = 0
        for b in range(64):
            hchk += (b + 1) * hist[b]
    print("MB09 mean=%s var=%s sd=%s min=%s max=%s mad=%s pearson=%s hist_chk=%d"
          % (fmt(mean), fmt(var), fmt(sd), fmt(mn), fmt(mx), fmt(mad), fmt(pear), hchk), flush=True)


def mb10():
    N, RR = 1000000, 3
    g = R(20270917)
    sum_v = 0; chk = 0; nbytes = 0; lines = 0
    for r in range(RR):
        name = "mb10_round_%d.txt" % r
        with open(name, "w", encoding="ascii") as f:
            for i in range(N):
                v = g.i()
                line = "%d %d\n" % (i, v)
                f.write(line)
                nbytes += len(line)
        with open(name, "r", encoding="ascii") as f:
            idx = 0
            for line in f:
                a, b = line.split(" ")
                ai = int(a); v = int(b)
                assert ai == idx
                idx += 1; lines += 1
                sum_v = (sum_v + v) % 1000000007
                chk = (chk * 31 + (v % 1000003)) % 1000003
    print("MB10 %d %d %d %d" % (sum_v, chk, nbytes, lines), flush=True)


def mb11():
    N, RR = 1000000, 3
    KM, SM, Q = 500009, 100003, 1000000007
    g = R(20271917)
    keys = [g.i() for _ in range(N)]
    acc = 0
    for r in range(RR):
        keys[r] = keys[r] + 1000000
        m = {}
        for i in range(N):
            k = keys[i] % KM
            m[k] = m.get(k, 0) + 1
        size1 = len(m)
        found = 0; vsum = 0
        for i in range(N):
            k = (keys[i] + 7) % KM
            if k in m:
                found += 1
                vsum += m[k]
        mchk = 0
        for k, v in m.items():
            bound((k % 1000003) * v)
            mchk = (mchk + (k % 1000003) * v) % 1000003
        for i in range(0, N, 2):
            k = keys[i] % KM
            if k in m:
                del m[k]
        size2 = len(m)
        st = set()
        for i in range(N):
            st.add(keys[i] % SM)
        size3 = len(st)
        lst = []
        for i in range(N):
            lst.append(keys[i] % 1000)
        lsum = 0
        for v in lst:
            lsum += v
        for v in (size1, found, vsum, mchk, size2, size3, lsum):
            bound(acc * 31 + v)
            acc = (acc * 31 + v) % Q
    print("MB11 %d %d %d %d %d %d %d %d" % (acc, size1, found, vsum, mchk, size2, size3, lsum), flush=True)


ALL = [mb01, mb02, mb03, mb04, mb05, mb06, mb07, mb08, mb09, mb10, mb11]
if __name__ == "__main__":
    want = sys.argv[1] if len(sys.argv) > 1 else "all"
    import time
    for f in ALL:
        if want == "all" or want == f.__name__:
            t0 = time.perf_counter()
            f()
            print("    # %s python %.2fs" % (f.__name__, time.perf_counter() - t0), flush=True)
    print("# max integer magnitude seen = %d (2^53 = %d) ok=%s" % (_bound, MAXEXACT, _bound < MAXEXACT), flush=True)
